imu_integrator/plugin.cpp

```diff
@@ -0,0 +1,254 @@
+	}
+
+private:
+	const std::shared_ptr<switchboard> sb;
+
+	// IMU Data, Sequence Flag, and State Vars Needed
+	std::unique_ptr<reader_latest<imu_cam_type>> _m_imu_cam;
+	std::unique_ptr<reader_latest<imu_integrator_seq>> _m_in;
+	std::unique_ptr<reader_latest<imu_integrator_input>> _m_imu_integrator_input;
+
+	// Write IMU Biases for PP
+	std::unique_ptr<writer<imu_raw_type>> _m_imu_raw;
+
+	std::vector<imu_type> _imu_vec;
+  	PimUniquePtr pim_ = NULL;
+
+	int counter = 0;
+	int cam_count = 0;
+	int total_imu = 0;
+	double last_cam_time = 0;
+	double last_imu_offset = 0;
+	long long _seq_expect, _stat_processed, _stat_missed;
+
+	// Open_VINS cleans IMU values older than 20 seconds, we clean values older than 5 seconds
+	void clean_imu_vec(double timestamp) {
+		auto it0 = _imu_vec.begin();
+        while (it0 != _imu_vec.end()) {
+            if (timestamp-(*it0).timestamp > 5) {
+                it0 = _imu_vec.erase(it0);
+            } else {
+                it0++;
+            }
+         }
+	}
+
+	// Timestamp we are propagating the biases to (new IMU reading time)
+	void propagate_imu_values(double timestamp, time_type real_time) {
+		const imu_integrator_input *input_values = _m_imu_integrator_input->get_latest_ro();
+		if (input_values == NULL) {
+			return;
+		}
+
+		ImuBias imu_bias = ImuBias(input_values->biasAcc, input_values->biasGyro);
+		if (pim_ == NULL) {
+			boost::shared_ptr<gtsam::PreintegratedCombinedMeasurements::Params> params =
+          			boost::make_shared<gtsam::PreintegratedCombinedMeasurements::Params>(input_values->params.n_gravity);
+
+			params->setGyroscopeCovariance(std::pow(input_values->params.gyro_noise, 2.0) * Eigen::Matrix3d::Identity());
+  			params->setAccelerometerCovariance(std::pow(input_values->params.acc_noise, 2.0) * Eigen::Matrix3d::Identity());
+  			params->setIntegrationCovariance(std::pow(input_values->params.imu_integration_sigma, 2.0) * Eigen::Matrix3d::Identity());
+			params->biasAccCovariance = std::pow(input_values->params.acc_walk, 2.0) * Eigen::Matrix3d::Identity();
+			params->biasOmegaCovariance = std::pow(input_values->params.gyro_walk, 2.0) * Eigen::Matrix3d::Identity();
+
+			pim_ = std::make_unique<gtsam::PreintegratedCombinedMeasurements>(params, imu_bias);
+			last_imu_offset = input_values->t_offset;
+		}
+
+		// Uncomment this for some helpful prints
+		// total_imu++;
+		// if (input_values->last_cam_integration_time > last_cam_time) {
+		// 	cam_count++;
+		// 	last_cam_time = input_values->last_cam_integration_time;
+		// 	std::cout << "Num IMUs recieved since last cam: " << counter << " Diff between new cam and latest IMU: " 
+		// 			  << timestamp - last_cam_time << " Expected IMUs recieved VS Actual: " << cam_count*10 << ", " << total_imu << std::endl;
+		// 	counter = 0;
+		// }
+		// counter++;
+		if (input_values->last_cam_integration_time > last_cam_time) {
+			std::cout << "&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&\n";
```

Comment:
Remove the couts before merging. You can guard them with `#ifndef NDEBUG` or remove them.